swarm/storage/localstore/gc.go

```diff
@@ -0,0 +1,184 @@
+	for {
+		select {
+		case <-db.collectGarbageTrigger:
+			// TODO: Add comment about done
+			collectedCount, done, err := db.collectGarbage()
+			if err != nil {
+				log.Error("localstore collect garbage", "err", err)
+			}
+			// check if another gc run is needed
+			if !done {
+				select {
+				case db.collectGarbageTrigger <- struct{}{}:
+				default:
+				}
+			}
+
+			if testHookCollectGarbage != nil {
+				testHookCollectGarbage(collectedCount)
+			}
+		case <-db.close:
+			return
+		}
+	}
+}
+
+// collectGarbage removes chunks from retrieval and other
+// indexes if maximal number of chunks in database is reached.
+// This function returns the number of removed chunks. If done
+// is false, another call to this function is needed to collect
+// the rest of the garbage as the batch size limit is reached.
+// This function is called in collectGarbageWorker.
+func (db *DB) collectGarbage() (collectedCount int64, done bool, err error) {
+	batch := new(leveldb.Batch)
+	target := db.gcTarget()
+
+	done = true
+	err = db.gcIndex.IterateAll(func(item shed.Item) (stop bool, err error) {
+		gcSize := atomic.LoadInt64(&db.gcSize)
+		if gcSize-collectedCount <= target {
+			return true, nil
+		}
+		// delete from retrieve, pull, gc
+		db.retrievalDataIndex.DeleteInBatch(batch, item)
+		db.retrievalAccessIndex.DeleteInBatch(batch, item)
+		db.pullIndex.DeleteInBatch(batch, item)
+		db.gcIndex.DeleteInBatch(batch, item)
+		collectedCount++
+		if collectedCount >= gcBatchSize {
+			// bach size limit reached,
+			// another gc run is needed
+			done = false
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return 0, false, err
+	}
+
+	err = db.shed.WriteBatch(batch)
+	if err != nil {
+		return 0, false, err
+	}
+	// batch is written, decrement gcSize
+	db.incGCSize(-collectedCount)
+	return collectedCount, done, nil
+}
+
+// gcTrigger retruns the absolute value for garbage collection
+// target value, calculated from db.capacity and gcTargetRatio.
+func (db *DB) gcTarget() (target int64) {
+	return int64(float64(db.capacity) * gcTargetRatio)
+}
+
+// incGCSize increments gcSize by the provided number.
+// If count is negative, it will decrement gcSize.
+func (db *DB) incGCSize(count int64) {
+	if count == 0 {
+		return
+	}
+	new := atomic.AddInt64(&db.gcSize, count)
+	select {
+	case db.writeGCSizeTrigger <- struct{}{}:
+	default:
+	}
+	if new >= db.capacity {
+		select {
+		case db.collectGarbageTrigger <- struct{}{}:
+		default:
+		}
+	}
+}
+
+var writeGCSizeDelay = 10 * time.Second
+
+// writeGCSizeWorker calls writeGCSize function
+// on writeGCSizeTrigger receive. It implements a
+// backoff with delay of writeGCSizeDelay duration
+// to avoid very frequent database operations.
+func (db *DB) writeGCSizeWorker() {
+	for {
+		select {
+		case <-db.writeGCSizeTrigger:
+			err := db.writeGCSize()
+			if err != nil {
+				log.Error("localstore write gc size", "err", err)
+			}
+			select {
```

Comment:
Is it okay if we got a trigger already but we don't write the gcSize? Should we write the gcSize as part of the shutdown?

Could you maybe also extract the wait `select` into a helper function? That kinda would help to keep the abstraction level constant in the function and definitely would have been easier to understand what this does. (backoff also confused me a little, as we are just rather doing a wait; backoff is more complicated in my mind like exponential backoff)

Reply:
Yes, you are right to write gcSize on DB Close. I added it.